[42,73,102,104,110] Modernization and refactoring

SwooshyCueb opened the pull request:
Addresses #42
Addresses #73
Addresses #102
Addresses #104
Addresses #110
In service of #95
In service of #103
Supersedes #96
Supersedes #97

Still a work in progress, but at a state where feedback is desirable.

This PR has a lot:
- Fix for the RPM transaction check error (#104)
- A little cleanup of `CMakeLists.txt`
- Use of `IrodsRunpathDefaults` in `CMakeLists.txt`
- Refactor to use nlohmann-json (#42)
- Refactor to use qpid-proton's C++ API (#73)
- Removal of json wrapper tokens (#102)
- `std::chrono` timestamp retrieval (in service of #95)
- Unstringified, `@timestamp`-keyed milliseconds-since-epoch timestamp (in service of #103)
- Unstringified pid in emitted json (in service of #103)
- Migration from `rodsLog` to `irods::experimental::log` (in service of #95)
- Removed unused `amqp_options` configuration setting (#110)
- Content type and creation timestamp in AMQP message metadata
- hostname from boost (in service of #95)
- better handling of default configuration
- Reformat to our new `clang-format` and `clang-tidy` configuration
- Migration from `boost::regex` to `std::regex` (in service of #95)
- Migration to `std::filesystem::path` for `test_mode` log (in service of #95)
- Removes old stomp-based python tests

trel commented:
if this confirms the new qpid-proton packages are good-to-go, i'll get them pushed to packages.irods.org

SwooshyCueb commented on .clang-format:
```diff
@@ -0,0 +1,203 @@
+---
```
I'm leaving this in for now, but it needs to be removed before the PR is ready for review

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,448 @@
+		return ERROR(SYS_LIBRARY_ERROR, fmt::format("[{}:{}] - [{}]", __func__, __LINE__, e.what()));
+	}
+	catch (const std::exception& e) {
+		return ERROR(SYS_INTERNAL_ERR, fmt::format("[{}:{}] - [{}]", __func__, __LINE__, e.what()));
```
`ERROR` already contains `__LINE__` and `__PRETTY_FUNCTION__`.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -208,7 +208,7 @@ auto start([[maybe_unused]] irods::default_re_ctx& re_ctx, const std::string& _i
 
 	try {
 		auto time_ms = ts_clock::now().time_since_epoch() / std::chrono::milliseconds(1);
```
should be `uint64_t` instead of `auto`

korydraughn replied:
Could replace this with std::chrono types instead of integer types too.

SwooshyCueb replied:
`uint64_t` was a deliberate choice, as it's the underlying type we want nlohmann-json to use.

korydraughn reviewed (changes requested):
So much better. I like the utc_clock reference.

korydraughn commented:
Let's replace the use of "Resolves" word in the first comment of this PR so that the issues aren't automatically closed.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,544 @@
+	}
+	catch (const irods::exception& e) {
+		// clang-format off
+		log_re::info({
```
`log_re::info` is the closest to the log-level that was used previously.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,544 @@
+	}
+}; // class send_handler
+
+static BOOST_FORCEINLINE void
```
The way to do this might differ between compilers. Boost provides a macro, so I used it.

korydraughn replied:
Sure. My question is more about what sparked you to add it? Was it there before or did you notice something that made you investigate it?

SwooshyCueb replied:
When I started working on this, these functions were neither static nor inline, even though they were not used outside of this translation unit.
Since the standard use case seems to be to have this plugin fire off for every rule that gets executed, I think inlining makes sense for the performance benefit.

korydraughn commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,544 @@
+	}
+	catch (const irods::exception& e) {
+		// clang-format off
+		log_re::info({
```
Oh. I hadn't realized that. I guess the question is should we switch to `log_re::error` here?

Let's create an issue for that. Your PR can address that in a separate commit if you want.

SwooshyCueb replied:
Ignore my previous comment. There *was* no exception handling previously to even be logged. I will make the suggested change.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,544 @@
+		arg_map.insert(std::make_pair(arg, 1));
+		return 1;
+	}
+	iter->second = iter->second + 1;
+	return iter->second;
```
Keeping this as-is. Not my code and I think doing it this way is clearer anyway.
To the uninitiated, `++iter->second` could be interpreted as `(++iter)->second` or `++(iter->second)`.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,413 +1,544 @@
+				// clang-format off
+				log_re::error({
+					{"rule_engine_plugin", rule_engine_name},
+					{"log_message", "unsupported argument for calling re rules from the rule language"},
```
`param` does not contain the name of the parameter, as far as I can tell. Even if it did, we wouldn't be able to trust it here since `serialize_parameter` failed.

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -267,7 +265,9 @@ auto start([[maybe_unused]] irods::default_re_ctx& re_ctx, const std::string& _i
 		json_obj["@timestamp"] = time_ms;
 
-		insert_or_parse_as_bin(json_obj, "hostname", host_name, time_ms);
+		std::string host_name = boost::asio::ip::host_name();
```
Should we consider using `boost::asio::ip::host_name()` in other places throughout iRODS, or is this specific to AMQP interactions?

SwooshyCueb replied:
`boost::asio::ip::host_name()` is a good C++-style replacement for `gethostname()` in most cases. So yeah I think it should be considered.

korydraughn replied:
Agreed

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -45,15 +45,21 @@ using ts_clock = std::chrono::system_clock;
+static const std::string rule_engine_name("audit_amqp");
+static const std::string default_pep_regex_to_match("audit_.*");
+static const std::string default_amqp_topic("irods_audit_messages");
+static const std::string default_amqp_location("localhost:5672");
```
Why not use brace-initialization?

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -193,20 +190,6 @@ auto get_re_configs(const std::string& _instance_name) -> irods::error
-						const auto& test_mode_str = test_mode_cfg->get_ref<const std::string&>();
-						test_mode = boost::iequals(test_mode_str, "true");
-						if (test_mode) {
-							const auto log_path_prefix_cfg = plugin_spec_cfg.find("log_path_prefix");
```
Do we want to remove the `log_path_prefix` configuration option? Just making sure we've thought that through. All the log information available through this facility is probably available through an AMQP log somewhere, right?

We should remove this from the test suite as well.

SwooshyCueb replied:
`log_path_prefix` was a `test_mode` thing. It serves no purpose without `test_mode`, which is removed in this PR (see #100).

alanking replied:
Understood. If we don't feel it is a useful concept even apart from `test_mode` then this is correct, yes.

SwooshyCueb replied:
Now that I'm actually looking at the stuff in `packaging/`, I think I'll leave the removal of `test_mode` to someone else.

alanking reviewed (commented):
Looking real good

SwooshyCueb commented:
Note to self:
somehow the logging modernization commit got squashed into the std::chrono commit

korydraughn commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,542 @@
+	class send_handler : public proton::messaging_handler
+	{
+	  private:
+		const std::string& amqp_location;
+		const std::string& amqp_topic;
+		const std::string& user;
+		const std::string& password;
+		const std::string& message;
```
Is it safe for these to be references?

SwooshyCueb replied:
Yes.
Aside from `message`, all of these come from the global configuration variables. The only thing that changes these is `get_re_configs()`, which is only called from `start()`.
At present, `send_handler` objects only live long enough to connect, send a single message, then disconnect. The only functions to construct `send_handler` objects are `start()`, `stop()`, and `exec_rule()`, all of which immediately lock `audit_plugin_mutex`.
Therefore, there is no possibility of the configuration variables being changed while a `send_handler` object is alive.

korydraughn replied:
I'm more so concerned with the following snippet leading to dangling references and potentially crashing the server.
```c++
send_handler handler(
￼  msg_str,
￼  audit_amqp_location,
￼  audit_amqp_topic,
￼  audit_amqp_user,
￼  audit_amqp_password);
proton::container(handler).run();
```
Is the message guaranteed to be sent by the time `.run()` returns?

SwooshyCueb replied:
To my understanding, yes.

korydraughn replied:
Good.

SwooshyCueb replied:
Just double-checked.
From the [API documentation for `proton::container::run()'](https://qpid.apache.org/releases/qpid-proton-0.36.0/proton/cpp/api/classproton_1_1container.html#a13a43e6d814de94978c515cb084873b1):
> Run the container in the current thread. 
> The call returns when the container stops.

korydraughn replied:
Awesome

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,534 @@
+#include <boost/config.hpp>
 #include <boost/regex.hpp>
 #include <boost/exception/all.hpp>
-#include <boost/algorithm/string.hpp>
```
don't remove

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -35,6 +34,17 @@
 
+// filesystem
+// clang-format off
+#ifdef __cpp_lib_filesystem
```
Should also conditionally add the boost library in cmake

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -38,14 +38,26 @@ namespace
+	std::string audit_amqp_options{default_amqp_options};
+
+	std::string log_path_prefix{default_log_path_prefix};
+	bool test_mode = default_test_mode;
```
Consider initializing the non-`const`s in `plugin_factory()`

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,607 @@
+			if (!password.empty()) {
+				conn_opts.password(password);
+			}
+			sender = container.open_sender(amqp_location + '/' + amqp_topic, conn_opts);
```
`fmt` might be better/faster

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,607 @@
 #include <boost/any.hpp>
+#include <boost/asio/ip/host_name.hpp>
+#include <boost/config.hpp>
 #include <boost/regex.hpp>
```
can probably be removed

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,607 @@
 #include <irods/irods_re_serialization.hpp>
 #include <irods/irods_server_properties.hpp>
 
 #undef LIST
```
figure out why this is needed and add a comment

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -123,32 +140,74 @@ namespace
 			}
-			sender =
-				container.open_sender(fmt::format(FMT_COMPILE("{0:s}/{1:s}"), amqp_location, amqp_topic), conn_opts);
+			container.open_sender(fmt::format(FMT_COMPILE("{0:s}/{1:s}"), amqp_location, amqp_topic), conn_opts);
```
location and topic combination can be done at config load

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -1,443 +1,471 @@
+			json_obj[key] = nlohmann::json::parse("\"" + val + "\"");
+		}
+		catch (const nlohmann::json::exception&) {
+			json_obj[key] = nlohmann::json::binary(std::vector<std::uint8_t>(val.begin(), val.end()));
```
This makes ES choke. Do something else instead.

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -55,6 +55,14 @@ namespace
 	// NOLINTEND(cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables)
 
+#if __cpp_lib_chrono >= 201907
+	// use utc_clock if it's implemented (for leap seconds)
```
This is important because AMQP considers/accounts for leap seconds?

SwooshyCueb replied:
We don't set AMQP-level timestamps. This is more for Elasticsearch, or whatever ends up consuming the AMQP messages. This is important because `system_clock` implements leap seconds by repeating a second. Since we use millisecond-level precision, this can result in messages appearing out of order when sorted by timestamp

alanking replied:
Okay. Perhaps a comment to that effect could be helpful for future readers?

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -224,7 +222,10 @@ namespace
 			insert_or_parse_as_bin(json_obj, "hostname", host_name);
-			json_obj["pid"] = std::to_string(pid);
+
+			pid_t pid = getpid();
+			json_obj["pid"] = pid;
```
Could we do like line 372 here?
```suggestion
			json_obj["pid"] = getpid();
```

SwooshyCueb replied:
No, because `pid` is used on line 407.

alanking replied:
Ah, the trouble with commit-level viewing. Got it.

alanking commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -280,7 +281,7 @@ namespace
 
 			pid_t pid = getpid();
-			json_obj["pid"] = std::to_string(pid);
+			json_obj["pid"] = pid;
```
Could we do like line 372 here?
```suggestion
			json_obj["pid"] = getpid();
```
Looks like this may be done in a later commit, but I'll leave this here just in case.

SwooshyCueb replied:
Yeah this is done in a later commit

alanking reviewed (commented):
I like the mechanism you built for the logging. Looks real nice.

Also, there's a typo in the commit message for 6234af996c702d8b472130303124fd44de042101: std::filesytstem::path -> std::filesystem::path

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -510,10 +506,7 @@ namespace
 
-			char host_name[MAX_NAME_LEN];
-			gethostname(host_name, MAX_NAME_LEN);
-			insert_or_parse_as_bin(json_obj, "hostname", host_name, time_ms);
```
Passing hostnames through `insert_or_parse_as_bin` was introduced in the qpid-cpp refactor commit. Instead of removing these calls here, just don't introduce them in the first place.

trel commented on src/amqp_sender.cpp:
```diff
@@ -0,0 +1,94 @@
+				{"rule_engine_plugin", rule_engine_name},
+				{"log_message", log_message},
+				{"error_condition::name", err_cond.name()},
+				{"error_condition::desc", err_cond.description()},
```
i see this is just 'moved code'... but is this a standard thing?   desc?    rather than description?

SwooshyCueb replied:
no, was introduced in the additional logging for proton commit. I abbreviated it for alignment with `name`, but I don't mind changing it.

trel replied:
i vote more letters unless there is a standing reason/standard/community that does it an existing way.  we live with these strings forever...

trel commented on README.md:
```diff
@@ -1,17 +1,17 @@
 # iRODS Rule Engine Plugin - Audit via AMQP 1.0
 
-This C++ plugin provides the iRODS platform a rule engine that can emit a single AMQP message to the configured topic for every policy enforcement point (PEP) encountered by the iRODS server.
+This C++ plugin provides the iRODS platform a rule engine that can emit a single AMQP 1.0 message to the configured topic for every policy enforcement point (PEP) encountered by the iRODS server.
```
I think it's still one-per-PEP, yes?

SwooshyCueb replied:
To my understanding, yes.

korydraughn replied:
I see. I think I glossed over the `... for every policy enforcement point (PEP) ...`.

SwooshyCueb commented on CMakeLists.txt:
```diff
@@ -115,16 +113,25 @@ install(
 
+add_subdirectory(unit_tests)
+
+set(PLUGIN_PACKAGE_NAME irods-rule-engine-plugin-hard-links)
```
good catch

alanking commented on src/main.cpp:
```diff
@@ -59,195 +62,39 @@ namespace fs = boost::filesystem;
 // clang-format on
 
-namespace
+namespace irods::audit
```
I think I like the direction this is headed (that is, making the plugin operations namespaced), but if we want to expand this pattern to other plugins, maybe we would want to establish a more general naming convention? Maybe something like:
```c++
namespace irods::plugin::rule_engine::audit_amqp
```
Thoughts? Too much?

SwooshyCueb replied:
I'm fine with that

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -50,6 +50,7 @@
 
 namespace
 {
+	const auto pep_regex_flavor = std::regex::ECMAScript;
```
Closest to what Boost uses

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -249,6 +252,8 @@ namespace
 		test_mode = default_test_mode;
 		log_path_prefix = default_log_path_prefix;
+
+		audit_pep_regex = std::regex(audit_pep_regex_to_match, pep_regex_flavor | std::regex::optimize);
```
Yeah the idea is that should the unthinkable happen, there will actually be something in `audit_pep_regex` no matter what

SwooshyCueb commented:
> How do we feel about main.cpp for the plugin filename? I think I'm okay with it, but just wanted to bring it up for discussion as we try to think about conventions in our plugins.

I'm not married to it. I can change it to `audit_amqp.cpp` if that is preferred.

korydraughn commented:
I think main.cpp is a very clear indicator of where someone should look when coming to an unfamiliar project. I prefer main.cpp, but am not against other ideas if they prove more helpful for us/others.

Logical quotas, metadata guard, and hard links REPs also use main.cpp too.

trel commented:
> Logical quotas, metadata guard, and hard links REPs also use main.cpp too.

seems main.cpp is both convention and expected

SwooshyCueb commented:
Addressed all review comments.
96bca7c01065085129357e13643b305024876881 might be of interest

korydraughn commented:
96bca7c looks fine overall. The only thing that stood out was the use of the C-style cast. A static_cast would be safer and better for tools. There may be an integer suffix for that too.

Other than that, was there anything in particular you feel we need to look at regarding that commit?

korydraughn commented:
Sadly, the integer suffix `zu` isn't available until C++23 (or with `-std=c++2b`).

SwooshyCueb commented:
Switched it to a `static_cast` and squashed. If there are no objections, I'll go ahead and \#

korydraughn commented:
Do all of the tests pass (python, cpp unit tests)?

SwooshyCueb commented:
cpp unit test suite passes. python test suites don't. looks like I've still got some work to do

SwooshyCueb commented:
Removed stomp-based python tests (no use updating since everything they covered is already covered by the proton-based python tests) and updated the proton-based python tests. All tests pass except one pulled from the unixfilesystem tests that appears unrelated.

SwooshyCueb commented on libirods_rule_engine_plugin-audit_amqp.cpp:
```diff
@@ -257,6 +254,19 @@ namespace
+						const auto amqp_options_cfg = plugin_spec_cfg.find("amqp_options");
+						if (amqp_options_cfg != plugin_spec_cfg.end()) {
+							// clang-format off
+							log_re::warn({
```
try to reduce the amount of spam that comes from this

SwooshyCueb commented:
If there are no objections, I will \# and cherry-pick to `4-3-stable`

korydraughn commented:
Is there one or more issues we can link the commits without an issue number to?

If not, let's create issues for them. Leave off the pounds so we can eyeball them one last time.

I'm assuming all of the tests pass now (I think you mentioned that this morning?).

trel commented:
yes, definitely need some squashing or issuing...

SwooshyCueb commented:
\#'d

korydraughn reviewed (approved):
From discussion this morning ...

Let's capture the failing unixfilesystem tests in an issue so we know to circle back to them.

alanking reviewed (commented):
Missing # on 41ff6daf344fa68bebd58f89754e47ff50835a10. Once that's added, will merge. And yes, please open an issue for the unixfilesystem test(s) which are failing.

So good

SwooshyCueb commented:
Fixed the missing \#, and added a change I forgot to include in the last commit (`postinst` script). Should be ready for merge now.
